WIP: Added Lupio-RTC device and new LupV platform

lhinman19 opened the pull request:
Change-Id: Ibd7cb3a36d9e1a897a20c462a90e04a0d9066a1a

joel-porquet commented on src/dev/lupio/LupioRTC.py:
```diff
@@ -0,0 +1,19 @@
+from m5.params import *
+from m5.proxy import *
+from m5.objects.Device import BasicPioDevice
+from m5.util.fdthelper import *
```
Double check that all of these imports are necessary.

powerjg replied:
To add, avoid `import *` if you can.

joel-porquet commented on src/dev/lupio/LupioRTC.py:
```diff
@@ -0,0 +1,19 @@
+    cxx_class='gem5::LupioRTC'
+    cxx_header = 'dev/lupio/lupio_rtc.hh'
+    time = Param.Time('01/01/2012',
+        "System time to use")
```
Does this need to be on two lines? It doesn't seem like the line would be over 80 characters.

joel-porquet commented on src/dev/lupio/LupioRTC.py:
```diff
@@ -0,0 +1,19 @@
+    type = 'LupioRTC'
+    cxx_class='gem5::LupioRTC'
+    cxx_header = 'dev/lupio/lupio_rtc.hh'
+    time = Param.Time('01/01/2012',
```
Is there a reason behind this date? Or is it just a copy from another file? What about the typical 01/01/1970 otherwise?

powerjg replied:
How about 1/1/2020 for gem5-20 (kinda)?

We probably don't want to use 1/1/1970 in case some installed program cares about the time. It may be weird to see the time be so old.

joel-porquet replied:
Funnily enough, I believe that raspberryPi don't have an RTC, so if they boot for the first time without network access to set up NTP, they actually show Jan 1st 1970!

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioRTC.hh"
+#include "sim/system.hh"
```
Double check these includes as well.

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+    gettimeofday(&tv, NULL);
+    gmtime_r(&tv.tv_sec, &time_bd);
+
+    DPRINTFN("Real-time clock set to %s", asctime(&time_bd));
```
I'm not sure what this constructor is supposed to do. Both variables are local, so they will disappear when the function returns.

powerjg replied:
Also on this line. Instead of `DPRINTFN`, it would be better to use `DPRINTF(LupioRTC, ...`

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+        case LUPIO_RTC_DYYR:
+            r = time_bd.tm_yday + 1;
+            break;
+    }
```
Make sure to keep the same comments as the QEMU model to explain the format of each field. Otherwise, it looks a bit random.

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+Tick
+LupioRTC::write(PacketPtr pkt)
+{
+    return pioDelay;
```
Is it the way to have the write operation be "unsupported"?

powerjg replied:
There's a couple of options here:

- Ignore writes like you do now.
- Print a warning `warn(...)` if you get a write to this device.
- Raise an error if you get a write: `panic("Unexpected write to the LupioRTC device!")`

I think the last is probably the right thing to do since this is for *teaching*. If a student writes a wrong driver this would catch it.

joel-porquet commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,59 @@
+    static const int LUPIO_RTC_CENT = 0x06;
+    static const int LUPIO_RTC_DYWK = 0x07;
+    static const int LUPIO_RTC_DYYR = 0x08;
+    static const int LUPIO_RTC_MAX = 0x09;
```
You could probably use an enum here.

joel-porquet commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,59 @@
+    /**
+    * Function to return current time and date using system's wall-clock time
+    */
+
```
Remove blank line between comment and function protype? (check with coding convention in gem5)

joel-porquet commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,59 @@
+    /**
+    * Implement BasicPioDevice virtual functions
+    */
+
```
Same here, can probably remove the blank line.

joel-porquet commented on src/dev/riscv/LupV.py:
```diff
@@ -0,0 +1,182 @@
+from m5.objects.Plic import Plic
+from m5.objects.RTC import RiscvRTC
+from m5.objects.Uart import Uart8250
+from m5.objects.Terminal import Terminal
```
Order imports alphabetically.

joel-porquet commented on src/dev/riscv/LupV.py:
```diff
@@ -0,0 +1,182 @@
+from m5.objects.Terminal import Terminal
+from m5.params import *
+from m5.proxy import *
+from m5.util.fdthelper import *
```
Again, double check that each import is necessary.

lhinman19 replied:
I double checked the imports and am now wondering if it is ok to have the `import *` for `params` and `fdthelper`?

powerjg replied:
Seems fine to me :)

joel-porquet reviewed (changes requested):
Great first PR! :)

I left a few comments for the code, and as mentioned in the chat, it'd be great to split the commit into two.

lhinman19 committed:
dev,arch-riscv: Introduced new LupV platform

This is a RISCV-based platform derived from the HiFive platform
containing the LupIO devices and will continue to be modified as more
devices are introduced. The LupIO devices are a collection of processor
agnostic and easily implemented virtual devices. Details about the
specifications of all eight LupIO devices can be found here:
https://gitlab.com/luplab/lupio/lupio-specs/-/tree/main

Information about how to build a RISCV full system with the LupIO-RTC
can be found here:
https://github.com/darchr/lupio-gem5/tree/configs

Change-Id: Ie15a217c9ab4e33d112acc3f498ba5b2eb4bb7d5

lhinman19 committed:
dev: Added new LupIO-RTC device

This device supplies the computer system with the simulated time by
retrieving the current tick number and converting to seconds, then
returning that time in ISO 8601 format. It is implemented as a
BasicPioDevice and is read-only.

The following are the specifications regarding the LupIO-RTC:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-rtc.md

Change-Id: Ic99291e5c862e728dac90b15ca818d42f25cbbee

powerjg commented on src/dev/lupio/LupioRTC.py:
```diff
@@ -0,0 +1,19 @@
+from m5.params import *
```
This file needs a license text. Copyright should be UC Regents. See src/arch/riscv/pmp.cc for an example.

powerjg commented on src/dev/lupio/SConscript:
```diff
@@ -0,0 +1,9 @@
+# -*- mode:python -*-
```
License. Also, remove the mode:python

powerjg commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+#include "dev/lupio/lupio_rtc.hh"
```
License :)

powerjg commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+    DPRINTFN("Real-time clock set to %s", asctime(&time_bd));
+}
+
+LupioRTC::~LupioRTC()
```
Don't need this function if it's empty. If it's pure virtual (which I don't think it is...) I *think* you can use `= default` in the header.

powerjg commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+    DPRINTF(LupioRTC, "time: %d\n", time_val);
+
+    pkt->setData(&time_val);
+    pkt->makeAtomicResponse();
```
Have you tested this in timing mode? I think this won't work. I believe you'll need to say:

```suggestions
    if(is_atomic) {
        pkt->makeAtomicResponse();
    } else {
        pkt->makeResponse();
    }
```

powerjg commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,103 @@
+    return pioDelay;
+}
+
+}
```
```suggestion
} // namespace gem5
```

It makes it easier to read :)

powerjg commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,59 @@
+
+using namespace RiscvISA;
+
+/** Real-Time Clock (LUPIO_RTC) */
```
Adding a more detailed doxygen comment would be great.

powerjg commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,59 @@
+    * Function to return current time and date using system's wall-clock time
+    */
+
+    uint8_t lupioRTCRead(const uint8_t addr);
```
This should probably be private. I don't think you would expect another object to call this function.

powerjg commented on src/dev/riscv/LupV.py:
```diff
@@ -0,0 +1,182 @@
+from m5.objects.Platform import Platform
```
License

powerjg commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,59 @@
+#ifndef __DEV_LUPIO_LUPIO_RTC_HH__
```
License :)

powerjg commented on src/dev/riscv/LupV.py:
```diff
@@ -0,0 +1,182 @@
+from m5.util.fdthelper import *
+
+class LupV(Platform):
+    """LupV Platform
```
Maybe a comment here saying that you copied this from HiFive.py and that you'll be updating it over a number of commits.

powerjg commented on src/dev/riscv/lupv.cc:
```diff
@@ -0,0 +1,62 @@
+#include "dev/riscv/lupv.hh"
```
License

powerjg commented on src/dev/riscv/lupv.hh:
```diff
@@ -0,0 +1,45 @@
+#ifndef __DEV_RISCV_LUPV_HH__
```
License

powerjg reviewed (commented):
There's probably a bit of a race condition here. Sorry!

One high-level thing... once we are able to drop the PLIC/CLINT we may want to revisit the need for the "platform" at all. I don't fully understand what that SimObject does...

powerjg commented:
For the commit messages, I would use "dev,riscv: ..." for the lupv platform and "dev:" for the RTC.

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,128 @@
+Tick
+LupioRTC::write(PacketPtr pkt)
+{
+    panic("Unexpected write to the LupioRTC device!");
```
Are you sure that panicking is the right way to go? Isn't there a more gentle way to go about it: like print a warning and/or return some error to the CPU as part of the simulation?

powerjg replied:
I think this is the right way to do it. There's two ways to look at it.
1. If the simulator has a bug and delivers a write to this device, we want to panic, it's a bug.
2. If a student is writing a device driver and does something wrong, rather than having to write good error handling code and checking the error correctly, I think we want to "fail fast" and have the simulator terminate. This is, in some sense, causing a "puff of smoke" out of the hardware.

powerjg replied:
A bus error would be a property of the bus protocol, right? What is the bus protocol for the RTC? It's just wires. There's no way to signal an error without having some kind of higher-level protocol in place. There's no default protocol in gem5 (or in any hardware). 

We can certainly just ignore writes, but that could mask bugs. I still think panic-ing (as defined by gem5) is the right thing to do. However, if you have an example where this device is written and it's not either a hardware or software bug, I can change my mind :)

joel-porquet replied:
Fine by me if it's the right way to go :+1:

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,128 @@
+    /**
+     * Get real time in seconds
+     */
+    gettimeofday(&tv, NULL);
```
I don't think your code compiles, does it? `tv` hasn't been declared.

joel-porquet commented on src/dev/lupio/lupio_rtc.cc:
```diff
@@ -0,0 +1,128 @@
+    gettimeofday(&tv, NULL);
+    gmtime_r(&tv.tv_sec, &time);
+
+    DPRINTF(LupioRTC, "Real-time clock set to %s", asctime(&time));
```
`tv` and `time` haven't been declared, and I'm still unclear on what this constructor is trying to do. Can you explain?

mkjost0 replied:
Currently, we have both `tv` and `time` defined in `lupio_rtc.hh`, because we use them in both the constructor and in LupioRTCRead.  In the constructor, we just wanted to grab the current time so we could use a debug flag to look at the time being saved while we were testing, since it was the only function that we were sure was being called while we still weren't finished.  However, we could now instead move this print statement to LupioRTCRead, make `tv` and `time` local variables to that function, and empty the constructor if that makes more sense to do?

joel-porquet replied:
Unless you really initialize the time in the constructor (in which case the device model needs to be somewhat rewritten to take that in account --but in that case, your variable should be part of the class itself), then you could just empty the constructor and the printf could be move to the read function (or even dropped altogether).

joel-porquet commented on src/dev/lupio/lupio_rtc.hh:
```diff
@@ -0,0 +1,93 @@
+
+        /**
+         * Max offset
+         */
```
For small comments like that, wouldn't `/* ... */` or `//` be enough? (confirm with coding conventions in gem5)

powerjg replied:
I think doxygen supports `///`

joel-porquet commented on src/dev/lupio/LupioRTC.py:
```diff
@@ -0,0 +1,43 @@
+    type = 'LupioRTC'
+    cxx_class='gem5::LupioRTC'
+    cxx_header = 'dev/lupio/lupio_rtc.hh'
+    time = Param.Time('01/01/2020', "System time to use")
```
This parameter doesn't seem to be actually used, since the RTC will recompute the time every time an access is made. Is there a reason to have it?

powerjg replied:
Let's chat about this tomorrow. I think we need to make a group decision on how to model the "time" in the RTC.

powerjg replied:
No, because that wouldn't be deterministic. It's important that every time you run a benchmark in gem5 that you get the same result.

If the user wants to use now(), they can do that from the config script easily.

joel-porquet replied:
Hmm, I forgot about the deterministic part. Ok, fine then.

powerjg reviewed (commented):
Looks great to me!

The final thing is the commit messages. The tags should probably be "dev:" for the one that adds the lupio RTC device and "dev,arch-riscv:" for the one that adds the lupv platform.

Then, you need to add enough info to the commit message so someone can review the commits. For instance, you need to include a link to the details of the LupIO devices (the specifications?) and some info about how you implemented it

powerjg commented on src/dev/lupio/SConscript:
```diff
@@ -0,0 +1,33 @@
+
+DebugFlag('LupioRTC')
+
+Source('lupio_rtc.cc')
```
That's not the gem5 style :). The file names match the class names 90 percent of the time.

joel-porquet replied:
Fine by me if it's the convention in gem5. But it *will* be a pain to navigate through files :P

joel-porquet commented:
Looks all good to me! I'll let you all handle the switch from `develop` to `lupio` for the branch to be merged in.

Happy Friday :smiley:
